src/library/ebuild-backend.c

```diff
@@ -0,0 +1,455 @@
+/**
+ * @file ebuild-backend.c
+ * @brief Implementation of the ebuild backend for fapolicyd.
+ *
+ * This file contains the implementation of the ebuild backend for fapolicyd.
+ * The ebuild backend is responsible for loading the list of installed packages
+ * and their corresponding files and directories from the VDB (/var/db/pkg/).
+ * It parses the CONTENTS file of each package and extracts the information
+ * about the installed files, including their paths, MD5 checksums, and modification timestamps.
+ *
+ * The ebuild_load_list function is the entry point for loading the package list.
+ * It takes a pointer to the conf_t structure, which contains the configuration options
+ * for fapolicyd, and returns an integer indicating the success or failure of the operation.
+ */
+
+// TODO: This needs to come from the build system
+#ifndef _DEFAULT_SOURCE
+#define _DEFAULT_SOURCE
+#endif
+#ifndef _POSIX_C_SOURCE
+#define _POSIX_C_SOURCE 200809L
+#endif
+
+#include <dirent.h>				// for dirent, closedir, opendir, DIR, readdir
+#include <errno.h>				// for errno
+#include <stdio.h>				// for NULL, perror, asprintf, getline, fopen
+#include <stdlib.h>				// for free, malloc, abort, reallocarray
+#include <string.h>				// for strcmp, strdup, strlen, strtok_r, strcat
+#include <syslog.h>				// for LOG_ERR, LOG_DEBUG, LOG_INFO
+#include "conf.h"				// for conf_t
+#include "config.h"				// for DEBUG
+#include "fapolicyd-backend.h"	// for SRC_EBUILD, backend
+#include "filter.h"				// for filter_destroy, filter_init, filter_l...
+#include "llist.h"				// for list_empty, list_init
+#include "md5-backend.h"		// for add_file_to_backend_by_md5
+#include "message.h"			// for msg
+
+static const char kEbuildBackend[] = "ebuilddb";
+
+static int ebuild_init_backend(void);
+static int ebuild_load_list(const conf_t *);
+static int ebuild_destroy_backend(void);
+
+backend ebuild_backend = {
+		kEbuildBackend,
+		ebuild_init_backend,
+		ebuild_load_list,
+		ebuild_destroy_backend,
+		/* list initialization */
+		{0, 0, NULL},
+};
+
+/*
+ * Collection of paths and MD5s for a package
+ */
+typedef struct contents {
+	char *md5;
+	char *path;
+} ebuildfiles;
```

Comment:
Nit, but this describes a single file, so it should properly be `ebuildfile` - otherwise it is confusing if the object is an array of file objects or just a single file.